qiskit_nature/properties/composite_property.py

```diff
@@ -0,0 +1,89 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A Composite of multiple properties."""
+
+from collections.abc import Iterable
+from typing import Dict, Generator, Generic, Optional, Type, TypeVar, Union
+
+from qiskit_nature.results import EigenstateResult
+from .property import Property
+
+# pylint: disable=invalid-name
+T = TypeVar("T", bound=Property)
+
+
+class CompositeProperty(Property, Iterable, Generic[T]):
```

Comment:
From [Wikipedia](https://en.wikipedia.org/wiki/Composite_pattern):
> The composite pattern describes a group of objects that are treated the same way as a single instance of the same type of object. The intent of a composite is to "compose" objects into tree structures to represent part-whole hierarchies.

I think in that sense, this is indeed a composite. However, as you also point out [below](https://github.com/Qiskit/qiskit-nature/pull/263#discussion_r669969445), it has the limitation of only being able to contain a single instance of any property type. I don't think that this is a limitation which will cause us issues, but if you disagree I can try to resolve that.

Reply:
To me this feels more like a set of properties - a PropertySet or PropertyGroup - rather than a composite. How do you composite spin and dipole together for any valid meaning at a conceptual level it just feels odd to me. Sure they can be grouped as a set so you can call interpret on the group and it iterates over each one; so it does have a method that acts as you describe on a result.

Reply:
I am fine with calling it a `GroupedProperty` or `PropertyGroup` if that is what you're after.

Reply:
Since it behaves like a single property with respect to the interpret acting on a single result maybe the former name `GroupedProperty` feels a bit more singular in that regard - that seemed to be the intent if I understand what you are looking for. I'd be ok with that too,

Reply:
Sounds good! Will update that tomorrow :+1: